openwisp_controller/config/who_is/service.py

```diff
@@ -0,0 +1,282 @@
+    @shared_task(
+        bind=True,
+        base=WhoIsCeleryRetryTask,
+        **app_settings.API_TASK_RETRY_OPTIONS,
+    )
+    def fetch_who_is_details(self, device_pk, initial_ip_address, new_ip_address):
+        """
+        Fetches the WhoIs details of the given IP address
+        and creates/updates the WhoIs record.
+        """
+        # The task can be triggered for same ip address multiple times
+        # so we need to return early if WhoIs is already created.
+        if WhoIsService._get_who_is_info_from_db(new_ip_address).exists():
+            return
+
+        WhoIsInfo = load_model("config", "WhoIsInfo")
+
+        ip_client = WhoIsService._get_geoip2_client()
+
+        try:
+            data = ip_client.city(ip_address=new_ip_address)
+
+        # Catching all possible exceptions raised by the geoip2 client
+        # logging the exceptions and raising them with appropriate messages
+        except (
+            errors.AddressNotFoundError,
+            errors.AuthenticationError,
+            errors.OutOfQueriesError,
+            errors.PermissionRequiredError,
+        ) as e:
+            exc_type = type(e)
+            message = EXCEPTION_MESSAGES.get(exc_type)
+            if exc_type is errors.AddressNotFoundError:
+                message = message.format(ip_address=new_ip_address)
+            raise exc_type(message)
+        except requests.RequestException as e:
+            raise e
+
+        else:
+            # Format address using the data from the geoip2 response
+            address = {
+                "city": data.city.name or "",
+                "country": data.country.name or "",
+                "continent": data.continent.name or "",
+                "postal": str(data.postal.code or ""),
+            }
+            # Create the WhoIs information
+            who_is_obj = WhoIsInfo(
+                isp=data.traits.autonomous_system_organization,
+                asn=data.traits.autonomous_system_number,
+                timezone=data.location.time_zone,
+                address=address,
+                cidr=data.traits.network,
+                ip_address=new_ip_address,
+            )
+            who_is_obj.full_clean()
+            who_is_obj.save()
+            logger.info(f"Successfully fetched WHOIS details for {new_ip_address}.")
+
+            # the following check ensures that for a case when device last_ip
+            # is not changed and there is no related who_is record, we do not
+            # delete the newly created record as both `initial_ip_address` and
+            # `new_ip_address` would be same for such case.
+            if initial_ip_address != new_ip_address:
+                # If any active devices are linked to the following record,
+                # then they will trigger this task and new record gets created
+                # with latest data.
+                WhoIsService.delete_who_is_record(ip_address=initial_ip_address)
```

Comment:
I can think of the following scenario where this could fail. Can you do a manual test for this? 

1. There are two devices from the same building. Hence, the `last_ip` of both the devices is same and they share a WhoIs record. 
2. One of the device was moved out of the building to somewhere else. It's last IP changes. This would delete Whois record for the old IP. 
3. The second device is working in the same building, and the `last_ip` of the device does not change. 

Will the logic create a new Whois record for the second device?

Reply:
The WhoIs record for the second device will be created when the device's checksum is requested. 
This was done to ensure we have latest details and any active device will create a WhoIs record (if it doesn't exists) when hitting the checksum endpoint.
This logic might be modified in the future when working on #1058 where we might have to add checks based on a threshold period using created_at and updated_at fields.